File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
from collections import defaultdict

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        f = defaultdict(int)
        max_f = 0
        l = r = 0
        while r < len(s):
            f[s[r]] += 1
            max_f = max(max_f, f[s[r]])
            if r - l + 1 - max_f > k:
                f[s[l]] -= 1
                l += 1
            ans = r - l + 1
            r += 1
        return ans

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        Map<Character, Integer> window = new HashMap<>();
        int maxFreq = 0;
        int res = 0;

        int start = 0;
        for (int end = 0; end < s.length(); end++) {
            char c = s.charAt(end);
            window.merge(c, 1, Integer::sum);
            maxFreq = Math.max(maxFreq, window.get(c));

            while (end - start + 1 - maxFreq > k) {
                char d = s.charAt(start);
                window.merge(d, -1, Integer::sum);
                start++;
            }

            res = Math.max(res, end - start + 1);
        }

        return res;
    }
}